### ISSUE TITLE
SMS MO / Encoding issue

### ISSUE THREAD

**Issue by cl77** [kneodev/ksmppd#166]
Hi there,

we have a little issue that clients do receive sms mo UTF-8 encoded.

We have created a dummy SMSC just for sms mo:

```
group = smsc
smsc = http
smsc-id = INJECT_SMS_MO
system-type = kannel
smsc-username = sms_mo
smsc-password = .....
port = 14000
send-url = "http://localhost:1234567"
log-level=0
log-file="/var/log/kannel/inject_sms_mo.log"
allowed-smsc-id=INJECT_SMS_MO
```

Then we re-route all sms mo to ksmppd:

```
group = smsbox-route
smsbox-id = ksmppd
smsc-id = INJECT_SMS_MO
```

KSMPPD then calls route.php to get the client who should receive the sms mo:

```
group=smpp-routing
routing-method=2
http-routing-url="https://our-domain.com/route.php"
http-num-callers=3
http-max-outstanding-requests=512
```

Works like a charm, a curl request to http://localhost:14000/cgi-bin/sendsms?... creates a sms mo which gets forwarded to client. E.G we send a SMS with coding=0 and content "€" thats what we receive: `data: 1b 65` ("€" in gsm7).

But if we do send "€" with coding=2 we receive UTF-8 encoded content: `data: e2 82 ac` (would have expected `20 ac`)
Same happens for concatenated gsm7 SMS. 

We have also tried to write sms mo entries directly to smpp_store table with similar results. I think we tried every combination of coding, charset etc.

```sql
insert into ksmppd.smpp_store 
    set 
      global_id = null, 
      sender = "4917622329545", 
      receiver = "4915735987970", 
      udhdata = null, 
      msgdata = "€ Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam.", 
      time = 1738591528, 
      smsc_id = "INJECT_SMS_MO", 
      smsc_number = null, 
      foreign_id = null, 
      service = "sms77test", 
      account = null, 
      id = "377b4300-a3d3-4c43-b3d7-6eb04a9d00c4", 
      sms_type = "0", 
      mclass = "-1", 
      mwi = "-1", 
      coding = "gsm", 
      compress = "-1", 
      validity = "-1", 
      deferred = "-1", 
      dlr_mask = "-1", 
      dlr_url = null, 
      pid = "-1", 
      alt_dcs = "-1", 
      rpi = "-1", 
      charset = "utf-8", 
      boxc_id = null, 
      binfo = null, 
      msg_left = "-1", 
      priority = "-1", 
      resend_try = "-1", 
      resend_time = "-1", 
      meta_data = null;
```

We're stuck now. Thanks in advance for any help!

BR  
CL

**Comment by mquevedob** [comment]
hello!

does this allow you to send a deliver sm to the ESME connected to the ksmppd ?

**Comment by cl77** [comment]
@mquevedob Exactly

**Comment by mquevedob** [comment]
@cl77 would it be too much to ask for you to show your ksmppd conf file and kannel conf file ?

**Comment by mquevedob** [comment]
@cl77 i keep getting Authorization failure when trying to call the sendsms url 

http://localhost:20002/cgi-bin/sendsms?user=sms_mo&pass=password&from=1234&to=5678&text=hola

my config is like this

group = smsc
smsc = http
smsc-id = loop1
system-type = kannel
smsc-username = sms_mo
smsc-password = password
port = 20002
send-url = "http://localhost:1234567"
log-level=0
log-file="/var/log/kannel/inject_sms_mo.log"
allowed-smsc-id=loop1

if you could give me some pointers i would be grateful

**Comment by cl77** [comment]
Weird. Did you setup routing in ksmppd? 

Our route.php (see above) simply returns some headers..

```
X-KSMPPD-Service: $smpp_username
X-KSMPPD-Route-Status: 1
X-KSMPPD-Route-Cost: 0
```

What does your log say when auth fails? And did you setup `group = smsbox-route` as above?

**Comment by mquevedob** [comment]
Are you sure you used smsc port in your sendsms url call ? Port 14000 in your example?

**Comment by cl77** [comment]
Yes, absolutely sure. Which version you use? I have 1.4.5

**Comment by mquevedob** [comment]
Same versión here

Can you share the complete url for sendsms endpoint ?

**Comment by cl77** [comment]
http://localhost:14000/cgi-bin/sendsms?username=sms_mo&password=...&to=....&from=...&charset=utf-8&coding=0&text=...

In case of ucs2 charset you have to convert encoding and set coding to 2.

**Comment by mquevedob** [comment]
would it be too much to ask for your kannel and ksmppd conf files ?

**Comment by mquevedob** [comment]
would it be too much to ask for your kannel and ksmppd conf files ?

**Comment by mquevedob** [comment]
hi!

i wanted to let everyone know that i made it work but not sending the /cgi-bin/sendsms request to the http SMSC port but to the sendsms-port of the smsbox group

thanks for all the support!

**Comment by cl77** [comment]
Interesting, i cant manage to get a SMS to ksmppd client with smsbox port. 

HTTP-based relay and content gateways should be exactly meant for uses like this:
https://www.kannel.org/download/kannel-userguide-snapshot/userguide.html#AEN2740

Could you please try to send a SMS MO to smpp client with €-sign? What do you get?

**Comment by cl77** [comment]
And did you also set a smsbox-route?

```
group = smsbox-route
smsbox-id = ksmppd
smsc-id = INJECT_SMS_MO
```

**Comment by mquevedob** [comment]
@cl77 yes, i added the smsbox-route group as well. it is working as expected

**Comment by cl77** [comment]
Could you please try to send this text?

> € Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam.

What do you receive at ksmppd client end? Is it `e2 82 ac` or `1b 65` for the €-sign?

**Comment by cl77** [comment]
Or maybe just add any UDH like `&udh=%05%00%03%cc%03%00` to sendsms url. Same result, i do reecive utf8 encoded content with warnings..

> 2025-02-03 11:46:51 [964628] [11] WARNING: Could not convert GSM (0xe2) to Unicode.
> 2025-02-03 11:46:51 [964628] [11] WARNING: Could not convert GSM (0x82) to Unicode.
> 2025-02-03 11:46:51 [964628] [11] WARNING: Could not convert GSM (0xac) to Unicode.


:-(
